data/modifications/pick_advance/dialog.lua

```diff
@@ -0,0 +1,168 @@
+	local recruits_subbutton = T.button { return_value = all_units_code, label = "\nSave (all)\n", tooltip = "Save the advancement for all units of this type" }
+	local recruits_button = not unit.canrecruit and T.row { T.column { horizontal_grow = true, recruits_subbutton } }
+
+-- main dialog definition
+	local dialog = {
+		T.tooltip { id = "tooltip_large" },
+		T.helptip { id = "tooltip_large" },
+		T.grid(filter_false {
+			T.row { T.column { T.spacer { width = 250 } } },
+			description_row,
+			T.row { T.column { horizontal_grow = true, listbox } },
+			T.row { T.column { horizontal_grow = true, unit_button } },
+			recruits_button,
+			T.row { T.column { horizontal_grow = true, reset_help_buttons } },
+		})
+	}
+
+-- dialog preshow function
+	local function preshow()
+		for i, advance_type in ipairs(options) do
+			local text = "\n" .. advance_type.name
+			if advance_type.id == game_override_one then
+				text = text .. " (chosen, all)"
+			elseif advance_type.id == unit_override_one then
+				text = text .. " (chosen)"
+			end
+			text = text .. "  \n"
+			wesnoth.set_dialog_value(text, "the_list", i, "the_label")
+			local img = advance_type.__cfg.image
+			wesnoth.set_dialog_value(img or "misc/blank-hex.png", "the_list", i, "the_icon")
+		end
+
+		wesnoth.set_dialog_focus("the_list")
+
+		local function select()
+			local i = wesnoth.get_dialog_value "the_list"
+			if i > 0 then
+				local img = options[i].__cfg.image
+				wesnoth.set_dialog_value(img or "misc/blank-hex.png", "the_list", i, "the_icon")
+			end
+		end
+		wesnoth.set_dialog_callback(select, "the_list")
+	end
+
+-- dialog postshow function
+	local item_result
+	local function postshow()
+		item_result = wesnoth.get_dialog_value("the_list")
+	end
+
+	local dialog_exit_code
+	while true do
+		dialog_exit_code = wesnoth.show_dialog(dialog, preshow, postshow)
+		if dialog_exit_code == help_code then
+			wesnoth.wml_actions.message {
+				speaker = "narrator",
+				message = "Choosing <u><b>Save for unit</b></u> will make your unit always advance to said type, even if it levels up during an enemy’s turn."
+					.. "\n\n"
+					.. "<u><b>Save for game</b></u> will do the same, but for all your new recruits of the same type in the game.",
+				image = "misc/qmark.png~SCALE(200,200)"
+			}
+		else
+			break
+		end
+	end
+
+-- determine the choice made
+	local is_reset = dialog_exit_code == reset_code
+	local is_ok = dialog_exit_code >= single_unit_code and item_result >= 1
+	--print(string.format("Button %s pressed (%s). Item %s selected: %s", dialog_exit_code, is_ok and "ok" or "not ok", item_result, options[item_result].id))
+	local game_scope = dialog_exit_code == all_units_code
+	return {
+		is_unit_override = is_reset or is_ok,
+		unit_override = is_ok and options[item_result].id or is_reset and table.concat(unit_type_options, ","),
+
+		is_game_override = is_reset or game_scope,
+		game_override = game_scope and options[item_result].id or nil,
+	}
```

Comment:
! Important !
The current add-on was written with the intention of shrinking the possible advancement options into one (because otherwise wesnoth would randomize).

With the current PR (https://github.com/wesnoth/wesnoth/pull/5491), we could change the logic to *reorder* the possible advancement list so that user's choice comes first. The user would then still be able to manually pick an advance if it happens at their turn.

I've received comments from people that asked for the add-on to work this way already. I could try help doing this rewrite.

Reply:
That sounds like a good idea to me.

Reply:
@vgaming to clarify - was this something you were planning to look into in the near future (within the next week or so)?  Either way is fine - there's no reason a second PR can't be made for updating the implementation - but if not I'm probably going to go ahead and merge this with the current implementation.

Reply:
Hey, I'm struggling between really wanting to do it, and not having time for it at the moment.

To explain the scope at least:
* We need to make the add-on reorder the advancement list, not shrinking it. (Q: should we re-check if other addons added new stuff to units?)
* We probably could eliminate or change the menu options. Right now the possibilities/options are: "assign for this unit", "assign for this unit and all new recruits". With the deterministic changes in mind, however, this could always default to changing recruits as well. It could be the worded as something like "pick advance for Mage" and it would assign advance for all Mages unassigned yet, plus all new Mage recruits.

Reply:
Leaving it at the discretion of the PR author to change it or define the deadline. Both make sense.

Reply:
I think I'd favor going ahead and merging it as-is so it'd be available for 1.15.10's release; changing the implementation to make use of reordering the advancements list instead would definitely be nice to have, but it's also a lower priority than getting the modification mainlined.

> should we re-check if other addons added new stuff to units?

I don't think it'd be necessary, unless you have an idea for it?

> We probably could eliminate or change the menu options. Right now the possibilities/options are: "assign for this unit", "assign for this unit and all new recruits". With the deterministic changes in mind, however, this could always default to changing recruits as well. It could be the worded as something like "pick advance for Mage" and it would assign advance for all Mages unassigned yet, plus all new Mage recruits.

I think that changing the advancement for just a single unit is probably something people would want to do.  Assigning it to every unit of that type where they haven't made an explicit choice would also somewhat go against one of the reasons for making advancements non-random: if they like the default, they don't need to change the advancement.